### ISSUE TITLE
"Error: Maximum update depth exceeded" when attempting to render Authenticator component with React 18's createRoot API

### ISSUE THREAD

**Issue by chrisbonifacio** [aws-amplify/amplify-ui#1602]
### Before creating a new issue, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-ui/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md#bug-reports).

### On which framework/platform are you having an issue?

React

### Which UI component?

Authenticator

### How is your app built?

Create React App

### Please describe your bug.

Trying to render the `Authenticator` or `withAuthenticator` components with React 18's `createRoot` API results in errors being thrown.

This newer render API might also be the cause of the Authenticator not rendering properly on newer Next.js (>= v[12.1.1](https://github.com/vercel/next.js/releases/tag/v12.1.1)) apps built using `create-next-app`.

### What's the expected behaviour?

Authenticator components should render without errors

### Help us reproduce the bug!

1. `npx create-react-app my-app`
2. `cd my-app`
3. `npm i aws-amplify @aws-amplify/ui-react`
4. copy paste provided code snippets into respective files
5. `npm run start`
6. Authenticator may render for a few seconds, try interacting with it to trigger the errors

### Code Snippet

```js
// index.js

import React from "react";
import * as ReactDOMClient from "react-dom/client";
import App from "./App";
import { Amplify } from "aws-amplify";
import awsmobile from "./aws-exports";

import "./index.css";

Amplify.configure(awsmobile);

const container = document.getElementById("root");

const root = ReactDOMClient.createRoot(container);

root.render(<App />);
```

```js
// App.js

import { withAuthenticator } from "@aws-amplify/ui-react";

const App = () => {
  return <h1>Welcome</h1>;
};

export default withAuthenticator(App);
```

### Additional information and screenshots
<img width="1455" alt="Screen Shot 2022-03-29 at 5 38 22 PM" src="https://user-images.githubusercontent.com/16952986/160711715-c7f8de6a-b438-44bf-a11b-d4eb3d1caa40.png">

**Comment by ErikCH** [comment]
Thanks @chrisbonifacio ! I was able to reproduce this. We'll have to do some work to get our Authenticator working with React 18!

**Comment by sarkarology** [comment]
Facing the same issue here.

Basic next.js setup with the Auth code from amplify docs.

Do we have any tentative date for the fix?

**Comment by OlehObodin** [comment]
any resolving?

**Comment by wlee221** [comment]
Hi all, I'm onto this and have found the root cause. This is due to cascading re-renders inside the `Authenticator`, and is being worked on in #1619 and #1620.

**Comment by stocad** [comment]
I ran into this while building a demo application just now and was able to work around it pinning to `"@aws-amplify/ui-react": "1.2.4",`. I would prefer to use the newer UI, but this worked and for the moment unblocks my progress. If there's a more current version that is usable while this gets fixed, it would be nice to have more current options.

**Comment by reesscot** [comment]
@stocad, the current version of `@aws-amplify/ui-react` (2.13.0) works with React 17. React 18 just landed a few days ago and as you can see in this ticket, we are working through a few issues with React 18 support. Please feel free to try the following if it would work for you:

### Downgrade to React 17 from 18:

1. Install React 17
```
npm install react@17 react-dom@17
```
or using yarn:
```
yarn add react@17 react-dom@17
```

2. Go to your index.js file. If it isn't already using `ReactDOM.render` change it back to use this function (see _before_ code in the [following guide](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis))


What are you using to build your demo application? I may be able to provide more specific steps if you still need help.

**Comment by wlee221** [comment]
Hi All, I just merged my fixes and will be released later today. Will let you know again once it's published!

**Comment by wlee221** [comment]
This has been now published to `@aws-amplify/ui-react@2.14.1`. Can you update to the latest version and retry? 

Please let us know if the problem persists, and please open a new issue if you find other frictions with React 18!

**Comment by techandmedia** [comment]
> This has been now published to `@aws-amplify/ui-react@2.14.1`. Can you update to the latest version and retry?
> 
> Please let us know if the problem persists, and please open a new issue if you find other frictions with React 18!

I can confirmed that this works without error with the latest amplify version, thanks
https://github.com/techandmedia/next-amplify-react18/tree/react18

**Comment by ianzone** [comment]
> This has been now published to `@aws-amplify/ui-react@2.14.1`. Can you update to the latest version and retry?
> 
> Please let us know if the problem persists, and please open a new issue if you find other frictions with React 18!

Hi, I still meet this issue when I use chrome to auto-fill the username and password. My dependencies are
    "@aws-amplify/ui-react": "^2.15.0",
    "aws-amplify": "^4.3.19",
    "next": "12.1.4",
    "react": "^18.0.0",
    "react-dom": "^18.0.0"

https://user-images.githubusercontent.com/20043230/162231795-dc46c52c-d3a9-49df-ba3e-1c71b4545aea.mp4


Unhandled Runtime Error
Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.

Call Stack
checkForNestedUpdates
node_modules/react-dom/cjs/react-dom.development.js (27084:0)
scheduleUpdateOnFiber
node_modules/react-dom/cjs/react-dom.development.js (25266:0)
dispatchSetState
node_modules/react-dom/cjs/react-dom.development.js (17342:0)
checkForUpdates
node_modules/use-subscription/cjs/use-subscription.development.js (96:0)
Object.setCurrentValue
node_modules/@aws-amplify/ui-react/node_modules/@xstate/react/es/useSelector.js (30:0)
eval
node_modules/@aws-amplify/ui-react/node_modules/@xstate/react/es/useSelector.js (66:0)
commitHookEffectListMount
node_modules/react-dom/cjs/react-dom.development.js (22969:0)
commitLayoutEffectOnFiber
node_modules/react-dom/cjs/react-dom.development.js (23081:0)
commitLayoutMountEffects_complete
node_modules/react-dom/cjs/react-dom.development.js (24461:0)
commitLayoutEffects_begin
node_modules/react-dom/cjs/react-dom.development.js (24447:0)
commitLayoutEffects
node_modules/react-dom/cjs/react-dom.development.js (24385:0)
commitRootImpl
node_modules/react-dom/cjs/react-dom.development.js (26651:0)
commitRoot
node_modules/react-dom/cjs/react-dom.development.js (26517:0)
performSyncWorkOnRoot
node_modules/react-dom/cjs/react-dom.development.js (25956:0)
flushSyncCallbacks
node_modules/react-dom/cjs/react-dom.development.js (11982:0)
flushSync
node_modules/react-dom/cjs/react-dom.development.js (26040:0)
finishEventHandler
node_modules/react-dom/cjs/react-dom.development.js (3969:0)
batchedUpdates
node_modules/react-dom/cjs/react-dom.development.js (3987:0)
dispatchEventForPluginEventSystem
node_modules/react-dom/cjs/react-dom.development.js (9257:0)
dispatchEvent
node_modules/react-dom/cjs/react-dom.development.js (6435:0)
dispatchDiscreteEvent
node_modules/react-dom/cjs/react-dom.development.js (6410:0)

**Comment by ErikCH** [comment]
Hi @ianzone !

Sorry you're still having this problem.

Let's double check a few things. 

Can you run `yarn why @aws-amplify/ui-react` or if you're on npm  `npm list @aws-amplify/ui-react` ? Also can you post your package.json? Maybe there is another dependency causing issues...

Are you using MFA, or confirming the sign up any ways? Also have you tried deleting your `yarn.lock` file and your `node_modules` and trying a fresh clean `npm install` / `yarn install` ?

**Comment by chrisbonifacio** [comment]
I wasn't able to reproduce this using auto-fill in Chrome or Edge

```js
// from package.json
  "dependencies": {
    "@aws-amplify/ui-react": "^2.11.0",
    "@emotion/react": "^11.8.2",
    "aws-amplify": "^4.3.18",
    "next": "^12.0.0",
    "react": "^18.0.0",
    "react-dom": "^18.0.0"
  },
 ```

**Comment by ianzone** [comment]
> 

Hi Erik, thanks for your reply.

The output of `npm list @aws-amplify/ui-react` is @aws-amplify/ui-react@2.15.0

My all dependencies are
  "dependencies": {
    "@aws-amplify/ui-react": "^2.15.0",
    "@progress/kendo-react-animation": "^5.2.0",
    "@progress/kendo-react-dialogs": "^5.2.0",
    "@progress/kendo-react-form": "^5.2.0",
    "@progress/kendo-react-grid": "^5.2.0",
    "@progress/kendo-react-inputs": "^5.2.0",
    "@progress/kendo-react-labels": "^5.2.0",
    "@progress/kendo-react-notification": "^5.2.0",
    "aws-amplify": "^4.3.19",
    "clean-deep": "^3.4.0",
    "next": "12.1.4",
    "react": "^18.0.0",
    "react-dom": "^18.0.0"
  },
  "devDependencies": {
    "@types/react": "17.0.43",
    "eslint": "8.12.0",
    "eslint-config-next": "12.1.4",
    "typescript": "4.6.3"
  }

I've set authenticator apps MFA as optional in Cognito, but I haven't set it in the frontend yet.

I've disabled the Sign Up option because the users are added by admins.

I've done `rm -rf node_modules && rm -f package-lock.json && npm cache clean --force && npm install`

I check with Firefox and it works fine, my chrome version is 99.0.4844.82 (Official Build) (64-bit)

**Comment by ianzone** [comment]
Hi @ErikCH @calebpollman , sry for drawing your concern. I just removed the `formFields` and `components` from the Authenticator props, and it works fine, so I guess sth went wrong there. 

here is my code for _app.tsx

```
import { Authenticator } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';
import { Amplify } from 'aws-amplify';
import { Layout } from 'components';
import { components, formFields } from 'components/Login';
import { AppContext } from 'contexts';
import type { AppProps } from 'next/app';
import 'styles/globals.css';
import awsConfig from '../aws-exports';
Amplify.configure(awsConfig);

function App({ Component, pageProps }: AppProps) {
  return (
    <Authenticator hideSignUp={true} formFields={formFields} components={components}>
      {({ signOut, user }) => {
        return (
          <AppContext.Provider value={{ signOut, user }}>
            <Layout>
              <Component {...pageProps} />
            </Layout>
          </AppContext.Provider>
        );
      }}
    </Authenticator>
  );
}
export default App;
```

**Comment by ianzone** [comment]
> ```diff
> context => [context.toResetPassword]
> ```

It works, thank you very much!

**Comment by wlee221** [comment]
Cool! We'll update our documentation to match that.
